jina/optimizers/new_flow_runner.py

```diff
@@ -0,0 +1,167 @@
+import os
+from itertools import tee
+from pathlib import Path
+import shutil
+from ruamel.yaml import YAML
+
+from jina.flow import Flow
+from jina.helper import colored
+from jina.logging import default_logger as logger
+
+
+class FlowRunner:
+    def __init__(
+        self,
+        flow_yaml,
+        document_generator,
+        batch_size,
+        pod_dir,
+        task,  # this can be only index or search as it is used to call the flow API
+        callback=None,
+        env_yaml=None,
+        overwrite_workspace=False,
+    ):
+        self.flow_yaml = flow_yaml
+        self.document_generator = document_generator
+        self.batch_size = batch_size
+        self.pod_dir = Path(pod_dir)
+        self.task = task
+        self.callback = callback
+        self.env_yaml = env_yaml
+        self.overwrite_workspace = overwrite_workspace
+
+    @staticmethod
+    def clean_workdir(workspace):
+        if workspace.exists():
+            shutil.rmtree(workspace)
+            logger.warning(colored("Existing workspace deleted", "red"))
+            logger.warning(colored("WORKSPACE: " + str(workspace), "red"))
+
+    @staticmethod
+    def _replace_param(parameters, trial_parameters):
+        for section in ["with", "metas"]:
+            if section in parameters:
+                for param, val in parameters[section].items():
+                    val = str(val).lstrip("$")
+                    if val in trial_parameters:
+                        parameters[section][param] = trial_parameters[val]
+        return parameters
+
+    def _create_trial_pods(self, trial_dir, trial_parameters):
+        if self.pod_workspace.exists():
+            shutil.rmtree(self.pod_workspace)
+        shutil.copytree(self.pod_dir, self.pod_workspace)
+        yaml = YAML(typ="rt")
+        for file_path in self.pod_dir.glob("*.yml"):
+            parameters = yaml.load(file_path)
+            if "components" in parameters:
+                for i, component in enumerate(parameters["components"]):
+                    parameters["components"][i] = self._replace_param(
+                        component, trial_parameters
+                    )
+            parameters = self._replace_param(parameters, trial_parameters)
+            new_pod_file_path = self.pod_workspace / file_path.name
+            yaml.dump(parameters, open(new_pod_file_path, "w"))
```

Comment:
That is also the way, we should handle yaml files in the discoverer. Thanks for showcasing it.